File: src/main/java/com/github/ccguyka/showupdates/slack/SlackBody.java
package com.github.ccguyka.showupdates.slack;

import static java.util.stream.Collectors.joining;

import com.github.ccguyka.showupdates.objects.ArtifactUpdate;
import com.github.ccguyka.showupdates.objects.DependencyUpdates;
import com.github.ccguyka.showupdates.objects.ProjectUpdates;

class SlackBody {

    private ProjectUpdates projectUpdates;
    private String artifact;

    public SlackBody(String artifact) {
        this.artifact = artifact;
    }

    static SlackBody artifact(String artifact) {
        return new SlackBody(artifact);
    }

    SlackBody updates(ProjectUpdates projectUpdates) {
        this.projectUpdates = projectUpdates;

        return this;
    }

    String build() {
        StringBuilder text = new StringBuilder();
        text.append("*Updates for _");
        text.append(artifact);
        text.append("_*");
        text.append("\n");
        text.append(printUpdates("parent", projectUpdates.getParent()));
        text.append(printUpdates("dependency", projectUpdates.getDependency()));
        text.append(printUpdates("plugin", projectUpdates.getPlugin()));
        text.append(printUpdates("dependency management", projectUpdates.getDependencyManagement()));

        return text.toString();
    }

    private String printUpdates(String type, final DependencyUpdates updates) {
        StringBuilder text = new StringBuilder();
        if (!updates.getArtifacts().isEmpty()) {
            text.append("\n");
            text.append("*Available " + type + " updates*");
            text.append("\n");

            for (final ArtifactUpdate dependencyUpdate : updates.getArtifacts()) {
                text.append(getLineText(dependencyUpdate));
            }
        }
        return text.toString();
    }

    private String getLineText(final ArtifactUpdate updates) {
        final StringBuilder result = new StringBuilder();
        result.append("• ");
        result.append(updates.getName());
        result.append(" ... ");
        result.append(updates.getCurrent());
        result.append(" -> ");
        result.append(updates.getUpdates().stream().collect(joining(",")));
        result.append("\n");
        return result.toString();
    }
}

File: src/main/kotlin/com/github/ccguyka/showupdates/slack/SlackBody.kt
package com.github.ccguyka.showupdates.slack

import com.github.ccguyka.showupdates.objects.ArtifactUpdate
import com.github.ccguyka.showupdates.objects.DependencyUpdates
import com.github.ccguyka.showupdates.objects.ProjectUpdates
import java.util.stream.Collectors

internal class SlackBody(private val artifact: String?) {

    private var projectUpdates: ProjectUpdates? = null

    fun updates(projectUpdates: ProjectUpdates?): SlackBody {
        this.projectUpdates = projectUpdates
        return this
    }

    fun build(): String {
        val text = StringBuilder()
        text.append("*Updates for _")
        text.append(artifact)
        text.append("_*")
        text.append("\n")
        text.append(printUpdates("parent", projectUpdates!!.parent))
        text.append(printUpdates("dependency", projectUpdates!!.dependency))
        text.append(printUpdates("plugin", projectUpdates!!.plugin))
        text.append(printUpdates("dependency management", projectUpdates!!.dependencyManagement))
        return text.toString()
    }

    private fun printUpdates(type: String, updates: DependencyUpdates): String {
        val text = StringBuilder()
        if (updates.artifacts.isNotEmpty()) {
            text.append("\n")
            text.append("*Available $type updates*")
            text.append("\n")
            for (dependencyUpdate in updates.artifacts) {
                text.append(getLineText(dependencyUpdate))
            }
        }
        return text.toString()
    }

    private fun getLineText(updates: ArtifactUpdate): String {
        val result = StringBuilder()
        result.append("• ")
        result.append(updates.name)
        result.append(" ... ")
        result.append(updates.current)
        result.append(" -> ")
        result.append(updates.updates.stream().collect(Collectors.joining(", ")))
        result.append("\n")
        return result.toString()
    }

    companion object {
        @JvmStatic
        fun artifact(artifact: String?): SlackBody {
            return SlackBody(artifact)
        }
    }
}
